specifyweb/frontend/js_src/lib/components/AttachmentsBulkImport/AttachmentsUpload.tsx

```diff
@@ -0,0 +1,428 @@
+                  }}
+                >
+                  {wbText.upload()}
+                </Button.Fancy>
+              </>
+            }
+            header={attachmentsText.beginAttachmentUpload()}
+            onClose={() => {
+              handleUploadReMap(undefined);
+            }}
+          >
+            {attachmentsText.beginUploadDescription()}
+          </Dialog>
+        ) : (
+          <Dialog
+            buttons={
+              <Button.DialogClose>{commonText.close()}</Button.DialogClose>
+            }
+            header={attachmentsText.attachmentServerUnavailable()}
+            icon={dialogIcons.warning}
+            onClose={() => {
+              handleSync(undefined, false);
+              setTriedUpload('main');
+            }}
+          >
+            <p>{attachmentsText.attachmentServerUnavailable()}</p>
+          </Dialog>
+        ))}
+
+      {
+        /*
+         * If upload was confirmed, but dataset status hasn't been set to uploading,
+         * the uploader is validating, and generating tokens. Display loading screen
+         * in that case
+         */
+        upload === 'confirmed' && dataSet.status !== 'uploading' && (
+          <LoadingScreen />
+        )
+      }
+    </>
+  );
+}
+
+function UploadState({
+  workProgress,
+  workRef,
+  onStop: handleStop,
+  onCompletedWork: handleCompletedWork,
+  triggerNow,
+}: AttachmentWorkStateProps): JSX.Element | null {
+  return workProgress.type === 'safe' ? (
+    <Dialog
+      buttons={
+        <Button.Danger onClick={handleStop}>{wbText.stop()}</Button.Danger>
+      }
+      header={wbText.uploading()}
+      onClose={undefined}
+    >
+      {attachmentsText.filesUploaded({
+        uploaded: workProgress.uploaded,
+        total: workProgress.total,
+      })}
+      <Progress max={workProgress.total} value={workProgress.uploaded} />
+    </Dialog>
+  ) : workProgress.type === 'stopping' ? (
+    <Dialog buttons={undefined} header={wbText.aborting()} onClose={undefined}>
+      {wbText.aborting()}
+    </Dialog>
+  ) : workProgress.type === 'stopped' ? (
+    <Dialog
+      buttons={<Button.DialogClose>{commonText.close()}</Button.DialogClose>}
+      header={wbText.uploadCanceled()}
+      onClose={() => handleCompletedWork(workRef.current.mappedFiles)}
+    >
+      {wbText.uploadCanceledDescription()}
+    </Dialog>
+  ) : workProgress.type === 'interrupted' ? (
+    <Dialog
+      buttons={
+        <>
+          <Button.Danger onClick={handleStop}>{wbText.stop()}</Button.Danger>
+          <Button.Fancy onClick={triggerNow}>
+            {attachmentsText.tryNow()}
+          </Button.Fancy>
+        </>
+      }
+      header={attachmentsText.interrupted()}
+      onClose={undefined}
+    >
+      {attachmentsText.interruptedTime({
+        remainingTime: formatTime(workProgress.retryingIn),
+      })}
+    </Dialog>
+  ) : null;
+}
+
+async function uploadFileWrapped<KEY extends keyof Tables>(
```

Comment:
too many arguments. could you receive arguments as an object (like in react components) to make things less bug-prone?